File: typescript/resistor-color-trio/resistor-color-trio.ts
const COLORS = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white',
]

const colorCode = (color: string): number => COLORS.indexOf(color)

export const decodedResistorValue = ([tens, ones, power]: string[]): string => {
  let ohms = (colorCode(tens) * 10 + colorCode(ones)) * 10 ** colorCode(power)
  let prefix = ''
  if(ohms > 10**9) {
    ohms/= 10**9
    prefix = 'giga'
  } else if(ohms > 10**6) {
    ohms/= 10**6
    prefix = 'mega'
  } else if(ohms > 1000) {
    ohms/= 10**3
    prefix = 'kilo'
  }
  return `${ohms} ${prefix}ohms`
}

File: ruby/resistor-color-trio/resistor_color_trio.rb
class ResistorColorTrio
  def initialize(colors)
    @colors = colors
    @ohms = calculate_ohms
  end

  def label
    "Resistor value: #{format_ohms}"
  end

  private

  def calculate_ohms
    tens = COLOR_CODES[@colors[0]] * 10
    units = COLOR_CODES[@colors[1]]
    multiplier = 10 ** COLOR_CODES[@colors[2]]
    (tens + units) * multiplier
  end

  def format_ohms
    if @ohms < 1000
      "#{@ohms} ohms"
    else
      "#{@ohms / 1000} kiloohms"
    end
  end

  COLOR_CODES = {
    "black" => 0,
    "brown" => 1,
    "red" => 2,
    "orange" => 3,
    "yellow" => 4,
    "green" => 5,
    "blue" => 6,
    "violet" => 7,
    "grey" => 8,
    "white" => 9
  }.freeze
end
